pkg/vsphere/performance/vm_collector.go

```diff
@@ -0,0 +1,407 @@
+		mos = append(mos, mo)
+	}
+	vmc.mu.Unlock()
+
+	// get the sample..
+	sample, err := vmc.perfMgr.SampleByName(op.Context, spec, names, mos)
+	if err != nil {
+		op.Errorf("unable to get metric sample: %s", err)
+		return
+	}
+	// convert to metrics
+	result, err := vmc.perfMgr.ToMetricSeries(op.Context, sample)
+	if err != nil {
+		op.Errorf("unable to convert metric sample to metric series: %s", err)
+		return
+	}
+
+	for i := range result {
+		met := result[i]
+		sub, exists := vmc.subs[met.Entity]
+		if !exists {
+			// the subscription is no longer valid go to the next result
+			continue
+		}
+
+		for s := range met.SampleInfo {
+
+			metric := &VMMetrics{
+				CPU: CPUMetrics{
+					CPUs: []CPUUsage{},
+				},
+				Memory:     MemoryMetrics{},
+				SampleTime: met.SampleInfo[s].Timestamp,
+				Interval:   sampleInterval,
+			}
+
+			// the series will have values for each sample
+			for _, v := range met.Value {
+				// skip the aggregate metric (empty string) and any negative values
+				if v.Instance == "" && v.Name != "mem.active.average" || v.Value[s] < 0 {
+					continue
+				}
+
+				switch v.Name {
+				case "cpu.usagemhz.average":
+					// convert cpu instance to int
+					cpu, err := instanceID(op, sub.ID(), v.Instance)
+					if err != nil {
+						// skipping this instance
+						continue
+					}
+					// specific vCPU metric
+					vcpu := CPUUsage{
+						ID:       cpu,
+						MhzUsage: v.Value[s],
+					}
+					metric.CPU.CPUs = append(metric.CPU.CPUs, vcpu)
+				case "mem.active.average":
+					metric.Memory.Active = v.Value[s]
+				// DISK
+				case "virtualDisk.read.average":
+					disk := findDisk(metric, v.Instance)
+					// perfManager returns kbps -- convert to sum of bytes
+					sum := summation(v.Value[s]) * 1024
+					metric.Disks[disk].Read.Bytes = sum
+					metric.Disks[disk].Read.Kbps = v.Value[s]
+				case "virtualDisk.write.average":
+					disk := findDisk(metric, v.Instance)
+					// perfManager returns kbps -- convert to sum of bytes
+					sum := summation(v.Value[s]) * 1024
+					metric.Disks[disk].Write.Bytes = sum
+					metric.Disks[disk].Write.Kbps = v.Value[s]
+				case "virtualDisk.numberReadAveraged.average":
+					disk := findDisk(metric, v.Instance)
+					// sum of iop read average
+					metric.Disks[disk].Read.Op = summation(v.Value[s])
+					metric.Disks[disk].Read.Ops = v.Value[s]
+				case "virtualDisk.numberWriteAveraged.average":
+					disk := findDisk(metric, v.Instance)
+					// sum of iop write average
+					metric.Disks[disk].Write.Op = summation(v.Value[s])
+					metric.Disks[disk].Write.Ops = v.Value[s]
+				// NET
+				default:
+					name := sub.DeviceName(v.Instance)
+					// if we have a name gather the stats
+					if name != "" {
+						// get the network to update
+						net := findNetwork(metric, name)
+						switch v.Name {
+						case "net.bytesRx.average":
+							// perfManager returns kbps -- convert to sum of bytes
+							sum := summation(v.Value[s]) * 1024
```

Comment:
`kbps * 1024` translates to `bps` - is `Rx.Bytes` actually `Rx.BytesPerSecond` or is there an unstated assumption that the interval is 1s?